src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/ProvisionContext.java

```diff
@@ -81,42 +84,44 @@ public void buildStack(final CloudPlatform cloudPlatform, Long stackId, Map<Stri
                     }
-                    List<Future<List<Resource>>> futures = new ArrayList<>();
+                    List<Future<Boolean>> futures = new ArrayList<>();
                     for (int i = 0; i < stack.getNodeCount(); i++) {
                         final int index = i;
                         final Stack finalStack = stack;
-                        Future<List<Resource>> submit = resourceBuilderExecutor.submit(new Callable<List<Resource>>() {
+                        Future<Boolean> submit = resourceBuilderExecutor.submit(new Callable<Boolean>() {
                             @Override
-                            public List<Resource> call() throws Exception {
+                            public Boolean call() throws Exception {
                                 LOGGER.info("Node {}. creation starting", index);
                                 List<Resource> resources = new ArrayList<>();
                                 for (final ResourceBuilder resourceBuilder : instanceResourceBuilders.get(cloudPlatform)) {
-                                    List<Resource> resourceList = resourceBuilder.create(pCO, index, resources);
-                                    stackUpdater.addStackResources(finalStack.getId(), resourceList);
-                                    resources.addAll(resourceList);
+                                    CreateResourceRequest createResourceRequest =
+                                            resourceBuilder.buildCreateRequest(pCO, resources, resourceBuilder.buildNames(pCO, index, resources), index);
+                                    stackUpdater.addStackResources(finalStack.getId(), createResourceRequest.getBuildableResources());
+                                    resourceBuilder.create(createResourceRequest);
+                                    resources.addAll(createResourceRequest.getBuildableResources());
                                     LOGGER.info("Node {}. creation in progress resource {} creation finished.", index, resourceBuilder.resourceBuilderType());
                                 }
-                                return resources;
+                                return true;
                             }
                         });
                         futures.add(submit);
                     }
-                    Exception exception = null;
-                    for (Future<List<Resource>> future : futures) {
+                    for (Future<Boolean> future : futures) {
                         try {
-                            resourceSet.addAll(future.get());
-                        } catch (Exception e) {
-                            exception = e;
+                            future.get();
+                        } catch (Exception ex) {
+                            throw new BuildStackFailureException(ex.getMessage(), ex, stackRepository.findOneWithLists(stackId).getResources());
                         }
                     }
```

Comment:
How is it ensured now that we won't lose any Exception messages? After the first Exception is noticed, the loop throws a failure and doesn't care about the other results.